Lux ai 2021

StoneT2000 opened the pull request:
Integrating the [Lux AI Season 1 Design](https://github.com/Lux-AI-Challenge/Lux-Design-2021/) into Kaggle Environments for a potential upcoming competition

harrisse commented:
FYI I'm going to start working my way through this and will try to help clean up any rough integration points with kaggle-environments as I go. Anything in particular you'd like me to focus on? Any parts of the review I should just skip (like boilerplate, 3rd party libraries, etc)?

harrisse commented:
Could you also provide instructions for how to run / test / debug this locally?

StoneT2000 commented:
@harrisse hi! So for game testing, we have our own test suites in a separate repository that holds our lux game design. Rest of this comment will bring up stuff on setup, skippable portions, and important changes

### Setup 

Run `kaggle_environments/envs/lux_ai_2021/dimensions/setup.sh`. This will install the game engine for lux ai and build the necessary code to run matches using kaggle-environments.

To test locally, I should have added some basic pytests to confirm it works. At the moment I use the following commands to generate either replay files that can be uploaded to a visualizer or generate an html file you can just open and watch

```
kaggle-environments run --environment lux_ai_2021 --agents js_agent_slow_expand random --render '{"mode": "html"}' --out replay.html 

kaggle-environments run --environment lux_ai_2021 --agents js_agent_slow_expand random --out out.json 
```

Agent ` js_agent_slow_expand` is a somewhat decent agent and is programmed using JS. We will have all competitors either write a python agent or write a JS or transpiled to JS agent (we will deal with transpilations ourselves). You can find a list of JS and python agents in `lux_ai_2021/test_agents` folder

### Parts you can probably skip:

Just about everything in the folder `kaggle_environments/envs/lux_ai_2021/dimensions`. There may be some edge cases where people submit bad bots that fail / crash (get error status) mid-way, at the beginning, or at the end, but I think these should be handled by the main environment file `kaggle_environments/envs/lux_ai_2021/lux_ai_2021.py`.

### Things to definitely look at:
`kaggle_environments/envs/lux_ai_2021/lux_ai_2021.py` - This has the code for interacting with our engine which is spawned as a separate process running in the background. Effectively kaggle-environments sends actions to the engine process, engine process returns observations. Python and JS agents parse these observations themselves in their own kits so while the replay data (the generated JSON) is basically unreadable, it is readable by the agents as they parse it themselves and store updated observations in a usable manner.

`kaggle_environments/utils.py` and `kaggle_environments/envs/lux_ai_2021/index.html` - Made some changes so that the `html_renderer` function can return a path to the html file which is our visualizer (which downloads assets and code from our website) and kaggle_environments can use that HTML, inject replay data, and it should work the same way the html render mode usually works.

Let me know if you have any questions

StoneT2000 commented:
Oh woop, since you last saw this, there has been a change. Updated the kaggle_env with the new design (we made some adjustments recently).

StoneT2000 commented:
Oh I think I may have made some accidental changes to the MANIFEST.in file and i don't think they do anything? I did add node_modules to gitignore as I assume we do not want to commit that mess of a folder

harrisse commented on kaggle_environments/envs/lux_ai_2021/dimensions/package-lock.json:
```diff
@@ -0,0 +1,5421 @@
+      "resolved": "https://registry.npmjs.org/dimensions-ai/-/dimensions-ai-4.8.0.tgz",
+      "integrity": "sha512-pC1CjeGEksILFEXmeHGJbpKX+oxfUjiZqNIJjvnPBHYlj8K2xlBOyPDGCHsf+1xaKC1PyG+/H9vq2bpiiw17FA==",
+      "dependencies": {
+        "@google-cloud/datastore": "^6.1.0",
```
Could you give me a rundown of your main dependencies and what they're used for? Are there any of these dependencies we can cut?

StoneT2000 replied:
Yes, I can actually trim the dependencies a lot.

The main package that runs our engine, `dimensions` has a few extra packages that should actually be peer dependencies or outright separate packages. This is something I will change as soon as I can.

StoneT2000 replied:
Ok actually smarter solution, got rid of all the `node_modules`, did some tree-shaking and packaged it all into a 1.7mb file :)

The main file is now in `kaggle_environments/envs/lux_ai_2021/dimensions/main.js`.

This is basically our game engine packaged neatly and is updated in our own private repo (will be made public upon release)

harrisse replied:
Apologies for the delay here, but nice approach! I'll take a look through and let you know if we need any further updates.

harrisse commented on kaggle_environments/envs/lux_ai_2021/index.html:
```diff
@@ -0,0 +1,42 @@
+    <script src="kaggle_environments/envs/lux_ai_2021/dist/vendors~app.js"></script>
+    <script src="kaggle_environments/envs/lux_ai_2021/dist/app.js"></script>
+    <script src="kaggle_environments/envs/lux_ai_2021/dist/phaser.js"></script> -->
+    <script src="https://2021vis.lux-ai.org/vendors~app~phaser.js"></script>
```
We'll need to be able to reference these locally rather than from an absolute URL.

StoneT2000 replied:
This is possible. I originally kept it as an absolute URL mostly because I didn't want to have to update kaggle envs every time the visualizer receives an update (that doesn't affect game logic). I will change this to be local instead.

Moreover, if I make it local, how do I know this will work in that embedded viewer on the kaggle website? I can get it working locally so that when render mode is HTML, it will produce a working HTML file but how will it work online?

Also maybe this is an indication that we should discuss about how this will be updated as the competition progresses? I expect there may be the occasional game logic or parameter change in the competition and updates to the visualizer.

StoneT2000 replied:
Actually follow up question, why do these need to be local? I see that in `static/player.html` there are files that are fetched from elsewhere:

https://github.com/Kaggle/kaggle-environments/blob/24a2cea763fe5d68a9f0f4c90b66d13801650316/kaggle_environments/static/player.html#L23

https://github.com/Kaggle/kaggle-environments/blob/24a2cea763fe5d68a9f0f4c90b66d13801650316/kaggle_environments/static/player.html#L51

Can we not do the same?

harrisse replied:
Could we obtain phaser through an official cdn? If so I'm fine with that.

harrisse reviewed (commented):
Checkpointing this review before the weekend, will pick up again on Monday :)

harrisse commented on docker/cpu.Dockerfile:
```diff
@@ -2,6 +2,42 @@
+RUN node -v
+RUN npm -v
+
+RUN npm install -g yarn typescript
```
(Please don't worry about doing this immediately, just curious if it's an option and if so how costly)

StoneT2000 replied:
Actually `yarn` is not necessary and we don't even use it. I think I can package the kaggle version of the lux engine so users will only need to run npm i lux-kaggle or whatever the package name is at the end. I will add this as soon as I can.

StoneT2000 replied:
and additionally then they don't need typescript to compile it since it'll already be compiled.

StoneT2000 replied:
Actually this thread is not relevant now, see https://github.com/Kaggle/kaggle-environments/pull/134#discussion_r651107550

StoneT2000 commented on docker/cpu.Dockerfile:
```diff
@@ -2,6 +2,42 @@
+RUN curl --silent -o- https://raw.githubusercontent.com/creationix/nvm/v0.31.2/install.sh | bash
+
+# install node and npm
+RUN . $NVM_DIR/nvm.sh \
```
I copied this from another pr from Bovard that wasn't merged. Feel free to change this however you like as long as we have access to typescript (tsc) and npm and nodejs

(sorry for the late comment, in exams right now and for some reason I was not given an email notification)

StoneT2000 commented on docker/cpu.Dockerfile:
```diff
@@ -2,6 +2,42 @@
+# update the repository sources list
+# and install dependencies
+RUN apt-get update \
+    && apt-get install -y curl \
```
feel free to remove

StoneT2000 commented on kaggle_environments/envs/lux_ai_2021/helpers.py:
```diff
@@ -0,0 +1,25 @@
+    @property
+    def last_opponent_action(self) -> int:
+        """Move the opponent took on the last turn."""
+        return self["halite"]
```
Actually i don't think these are necessary? The way raw observations are parsed are all handled in the kit folder for the python agent `kaggle_environments/envs/lux_ai_2021/test_agents/python/lux` (and JS agents have their own kit for parsing it).

Where are these helpers used?

harrisse replied:
They're just for users to make their own bots more easily, if you don't need them it was likely just a copy paste from an older env and they can probably be deleted.

StoneT2000 commented:
@harrisse do we have a decision on how users should install the game engine? I assume I can still let users just run the script I provided (minus the typescript compiling part, just direct npm install)

StoneT2000 commented:
> @harrisse do we have a decision on how users should install the game engine? I assume I can still let users just run the script I provided (minus the typescript compiling part, just direct npm install)

Actually ignore this now, user's no longer need to do any install, they just need nodejs

StoneT2000 commented:
# Updates

Going through what top level packages are packaged into the engine (which are the same top level packages people end up installing to run locally) and also summarizing changes since last time.

## Top Level Packages
- "colors": "^1.4.0", - colors in terminal
- "jszip": "^3.5.0", - compressing replays
- "nanoid": "^3.1.12", - generating unique ids
- "seedrandom": "^3.0.5", - seeded RNG (JS does not have seeded RNG)
- "yargs": "^17.0.1" - command line argument parser
- "dimensions-ai": "^5.0.1", - our own package for running matches locally and handling different languages, printing warnings (a bunch of stuff), making dev life easier in general
  - these are the top level packages that people would download when installing our game engine
    - "colors": "^1.4.0", - colored terminal messages
    - "cross-spawn": "^7.0.2", - spawning processes (for running the agents) cross platform (windows, mac, linux...)
    - "nanoid": "^3.1.4", - generating unique ids
    - "pidusage": "^2.0.18", - monitoring agent memory use
    - "process-exists": "^4.0.0",
    - "rimraf": "^3.0.2", - remove files/folders cross platform
    - "seedrandom": "^3.0.5", - RNG with seeding (Javascript does not have seeded RNG)
    - "sprintf-js": "^1.1.2", - pretty printing messages
    - "tree-kill": "^1.2.2",
  - these are top level but are basically tree shaken out unless someone chooses to install the full dimensions package to gain access to some other (not necessary) features:
    "dockerode": "^3.2.1", (for docker stuff but via node)
    "ts-trueskill": "^2.5.5", (true skill ranking)
    "express": "^4.17.1", (for running a local server to interact with the game engine, a lot of the following packages are for the server)
    "dotenv": "^8.2.0",
    "extract-zip": "^2.0.0",
    "formidable": "^1.2.2",
    "jsonwebtoken": "^8.5.1",
    
    "bcryptjs": "^2.4.3",
    "body-parser": "^1.19.0",
    "cors": "^2.8.5",

## Changes

not much actually, I think we got most of it covered in the meeting earlier. Just clarifying a few things so we're on the same page

Competitors can submit bots as long as they have a main.py file with a function that accepts observations (and the 2nd configs parameters). If they use python/nodejs, there is "tech support" from you all. Otherwise, we give no guarantees but they are free to submit as they like

Most of our competitors may run matches using our own packaged CLI tool for running them (has a # of helpful features). Some may also use kaggle-environments to run things. Either way, i've tested (locally) and there is 0 discrepancies  (it's the same engine after all). I will try my best to add some good pytests for this though (among other things probably)

## Action items

I'm meeting with bovard soon to have another discussion probably 

Otherwise action items for me include
- move viewer files to a cdn
- remove the old helpers.py code
- update engine code to latest version (we're change it a lot at the moment)

For you all probably just testing this out, and seeing if its mergeable yet

StoneT2000 commented:
Just wanted to clarify if unpkg works https://unpkg.com/ (saw one of the html files originally had it)

harrisse commented:
Yep, unpkg works.

StoneT2000 commented on kaggle_environments/envs/lux_ai_2021/test_lux.py:
```diff
@@ -0,0 +1,17 @@
+    assert json["statuses"] == ["DONE", "DONE"]
+
+def test_js_agents():
+    env = make("lux_ai_2021", configuration={})
```
fix, i removed organic agent earlier.

bovard commented on kaggle_environments/envs/lux_ai_2021/todo.md:
```diff
@@ -0,0 +1,18 @@
+# TODOs
```
Is this list current / which of these need to be done before we launch beta?

StoneT2000 replied:
this is all done with one exception. It would be nice to somehow retrieve the name of a team and store it into the replay file.  If it's not possible then it's fine, people will just need to know if theyre team or team 1.

bovard replied:
@harrisse is it possible to store team names in the replay?

The rest of these should be added as github issues rather than a todo here. This file can be removed.

harrisse replied:
Yes, for prod episodes you'll get team names in the replay under the top level info property, e.g.:

```"info": {"EpisodeId": 23575003, "LiveVideoPath": null, "TeamNames": ["Goose luck", "Goosebumps", "HandyRL", "NejuMixWATORI"]}```

StoneT2000 replied:
Ok this does not require a change to this PR. This will be a change on our side for the visualizer at https://github.com/Lux-AI-Challenge/LuxViewer2021. At the moment when loading kaggle replays, the visualizer just puts blanks for the names, to be changed soon.

bovard commented on kaggle_environments/agent.py:
```diff
@@ -34,15 +34,21 @@ def is_url(url):
 
     try:
-        code_object = compile(raw, "<string>", "exec")
+        code_object = compile(raw, path, "exec")
```
@harrisse more unfamiliar with this file, do these changes look good here?

harrisse replied:
Besides the configuration line below, yes, these are nice updates, thanks Stone :)

StoneT2000 commented on kaggle_environments/envs/lux_ai_2021/notes.md:
```diff
@@ -0,0 +1,30 @@
+RPS Lux (1k steps)
+python kaggle_environments/envs/rps/scratch.py  3.48s user 0.58s system 124% cpu 3.250 total
+
+RPS Python (1k steps)
```
removed

StoneT2000 commented on kaggle_environments/envs/lux_ai_2021/lux_ai_2021.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def renderer(state, env):
+    sign_names = ["Rock", "Paper", "Scissors", "Spock", "Lizard"]
```
should we raise an error if someone tries to do some kind of ansi output?

StoneT2000 replied:
ok at the moment it just raises an error and tells the user to use json or html rendering.

StoneT2000 commented:
@bovard this is another thing that I don't remember if this was discussed but it would be nice if we could change the observations field in the replay file. It's currently super super verbose and could be replaced by something that is much smaller or more verbose but more useful. Otherwise a game could take up like 17mb of space if not compressed.

I might also just be using the api wrong.

StoneT2000 commented on kaggle_environments/envs/lux_ai_2021/testing.md:
```diff
@@ -0,0 +1,9 @@
+Some tests can be automated, however since the actual design is programmed in a seperate repository and uses a different engine for testing, some things needs to be manually checked at the moment
+
+Whenever the design is changed, the following tests should be done:
+
```
There isn't a programmatic way at the moment because kaggle replay files are formatted differently than the replay files players generate locally (this can be alleviated later with some work, either change what is stored in the replay by editing the observations saved or write a converter). I do this by running a game in kaggle like so

```
kaggle-environments run --environment lux_ai_2021 --agents simple_agent simple_agent --render '{"mode": "json"}' --out out.json --debug=True
```

Then upload the out.json to https://2021vis.lux-ai.org/ and scroll to the final turn.

Then using our own engine (which you install via `npm i -g install @lux-ai/2021-challenge`), and download the simple kit from https://github.com/Lux-AI-Challenge/Lux-Design-2021/tree/master/kits/python (or another language)

and run

```
lux-ai-2021 path/to/simple/main.py path/to/simple/main.py --out=replay.json
```

upload to replay viewer again and compare the final state with the other kaggle produced match.

bovard replied:
Can you add those instructions to that file?

StoneT2000 replied:
yup!

harrisse commented on kaggle_environments/agent.py:
```diff
@@ -104,20 +110,20 @@ def build_agent(raw, builtin_agents, environment_name):
-            agent = get_last_callable(raw) or raw
-
+            agent = get_last_callable(raw_agent, path=raw) or raw_agent
+        configuration["__raw_path__"] = raw
```
Is this needed?

StoneT2000 replied:
this makes it so that people using our starter kits can run bots by passing a path to their main.py file (for wtv language they choose to write the bot in) and code their bot without having to worry about path shenanigans.

Here's an example in our C++ agent's main.py file: https://github.com/Lux-AI-Challenge/Lux-Design-2021/blob/master/kits/cpp/simple/main.py at line 19 we use it. We need the path so that the current working directory can be set for the running process (makes things easier for competitors if they know things are running in the root bot directory).

StoneT2000 commented on kaggle_environments/envs/lux_ai_2021/lux_ai_2021.py:
```diff
@@ -0,0 +1,120 @@
+
+    player1.observation.player = 0
+    player2.observation.player = 1
+
```
actually @harrisse are ties allowed? if so I'd rather we allow ties (our last tie breaker is a coin flip)

bovard reviewed (approved):
Couple more small changes, then LGTM

StoneT2000 commented:
Awesome!
